test/pytorch_test_base.py

```diff
@@ -390,39 +390,49 @@ def disallowed_test(self, test=test):
         super().instantiate_test(name, test)
       else:  # Tests with dtype variants have unsupported dtypes skipped
         # Sets default precision for floating types to bfloat16 precision
         if not hasattr(test, 'precision_overrides'):
           test.precision_overrides = {}
         xla_dtypes = []
-        for dtype in dtypes:
-          dtype_str = str(dtype).split('.')[1]
-          dtype_test_name = test_name + '_' + dtype_str
-          if dtype in cls.unsupported_dtypes:
-            reason = 'XLA does not support dtype {0}'.format(str(dtype))
-
-            @wraps(test)
-            def skipped_test(self, *args, reason=reason, **kwargs):
-              raise unittest.SkipTest(reason)
-
-            assert not hasattr(
-                cls, dtype_test_name), 'Redefinition of test {0}'.format(
-                    dtype_test_name)
-            setattr(cls, dtype_test_name, skipped_test)
-          elif match_name(dtype_test_name, disabled_torch_tests[class_name]):
-            setattr(cls, dtype_test_name, disallowed_test)
-          else:
-            xla_dtypes.append(dtype)
-          if dtype in [torch.float, torch.double, torch.bfloat16]:
-            floating_precision = XLATestBase._alt_lookup(
-                TORCH_TEST_PRECIIONS,
-                [dtype_test_name, test_name, test.__name__],
-                DEFAULT_FLOATING_PRECISION)
-            test.precision_overrides[dtype] = floating_precision
-
+        for dtype_combination in dtype_combinations:
+          if type(dtype_combination) == torch.dtype:
+            dtype_combination = (dtype_combination,)
+          dtype_test_name = test_name
+          skipped = False
+          for dtype in dtype_combination:
+            dtype_test_name += '_' + str(dtype).split('.')[1]
+          for dtype in dtype_combination:
+            if dtype in cls.unsupported_dtypes:
+              reason = 'XLA does not support dtype {0}'.format(str(dtype))
+
+              @wraps(test)
+              def skipped_test(self, *args, reason=reason, **kwargs):
+                raise unittest.SkipTest(reason)
+
+              assert not hasattr(
+                  cls, dtype_test_name), 'Redefinition of test {0}'.format(
+                      dtype_test_name)
+              setattr(cls, dtype_test_name, skipped_test)
+              skipped = True
+              break
+            if dtype in [torch.float, torch.double, torch.bfloat16]:
+              floating_precision = XLATestBase._alt_lookup(
+                  TORCH_TEST_PRECIIONS,
+                  [dtype_test_name, test_name, test.__name__],
+                  DEFAULT_FLOATING_PRECISION)
+              test.precision_overrides[dtype] = floating_precision
+
+        if match_name(dtype_test_name, disabled_torch_tests[class_name]):
+          skipped = True
+          setattr(cls, dtype_test_name, disallowed_test)
+        if not skipped:
```

Comment:
I am not sure this is correct, if I understood the issue.
If a test is passed (int32, bfloat16, float32) and bfloat16 is not wanted, we should run the test with (int32, float32).
This was the original intent IIRC.
Now you say we are passed things like (int32, (int64, bfloat16), float32)?
Is this what the PR is trying to fix?
In that case we still want to run the (int32, float32) tests, which this PR seems to prevent.

Reply:
I think I messed up the indent here, these checks should be within the `for dtype_combination in dtype_combinations` loop.

For tests like `test_result_type`, it will get expanded to `test_result_type_xla_float16_float16`. This kind of test takes 2 dtype for each instance of test. In `dtype_combinations`, we will have a list of tuple of dtypes 

```
((torch.bool, torch.bool), (torch.bool, torch.uint8), (torch.bool, torch.int8), (torch.bool, torch.int16), (torch.bool, torch.int32), (torch.bool, torch.int64), (torch.bool, torch.float16)......)
```

so `dtype_combination` in above code would be either a tuple of dtypes like`(torch.bool, torch.int8)` or a single dtype like `torch.float16`

The intent of this change is to handle the dtype tuple case. If either of the dtyps is a `cls.unsupported_dtypes`, we should skip the test.